### ISSUE TITLE
Specs: Utility pack.

### ISSUE THREAD

**Issue by frandiox** [OnsenUI/OnsenUI#1447]
@argelius @anatoo @philolo1 @IliaSky @masahirotanaka 
Some love for Vanilla JS users. These are just ideas with a possible syntax, but opinions/suggestions/modifications are welcome.
- Templating: Possibility of passing a function to navigation methods (pushPage, load, etc) that will be run before the content of `ons-template`/file is used by `getPageHTMLAsync` or similars. Users will be able to have real templates since they can plug handlebars or anything in that function. Perhaps it would be also interesting to set a function for a specific template so they don't need to pass the same function in every `pushPage`.

```
// First way
navi.pushPage(page1.html,
  { 
    data: { title: 'Page 1' }, 
    templating: function(templateString, data) {
      return templateString.replace('{{title}}', data.title);
    }
  }
);

// Second way
<ons-template id="page1.html" templating="app.f1"> ... </ons-template>
```
- `ons-include`: Includes the content of any `ons-template`/file with the previously explained templating function. This means that `ons-template` could contain anything, not only pages.

```
<ons-include source="toolbar.html" templating="app.fToolbar1"></ons-include>
```
- `ons-repeat`: Similar to lazy-repeat component but without unloading elements.

```
<ons-list>
  <ons-repeat delegate="app.itemConstructor">
    <ons-list-item>{{content}}</ons-list-item>
  </ons-repeat>
</ons-list>

///

app.delegate = {
  createItemContent: function(index, templateString) {
    return templateString.replace('{{content}}', myData[index].content);
  },
  countItems: myData.length
}
```
- `ons.util.createElement` or `ons.createElement`: If this is the only utility function we want to make public perhaps we can put it directly inside `ons`: `var listItem = ons.createElement('<ons-list-item modifier="longdivider"><div class="center">...</div></ons-list-item>');`

**Comment by munsterlander** [comment]
YES PLEASE!!!  I would love include and repeat especially if I can make it play with VueJS.  Create element would be my alternate to repeat because I can loop and use that, which technically I already am doing, but I digress.

**Comment by IliaSky** [comment]
I guess I already said what's my opinion on the subject irl, but lets log it here too:

- Templating - totally up for it. When I joined the team I also thought it was missing and that a template without a templating function doesn't feel right. And for the implementation - I think I would actually prefer a way to set it globally (something like `ons.setTemplatingEngine()`). As for the suggested ways - the second one in theory sounds easier, maybe both would be fine though. We should try to unify the implementation for creating pages though, instead of repeating the same logic in many places.
- `ons-include` - while in theory this can worsen the performance if used incorrectly I still like it - again it sounds useful and probably a lot of people would like it. If we decide to include non-pages in the template I guess we could make the auto-page wrapping in the `loadPage` utility function or whatever it will be. I think removing that feature may break a lot of things and also I don't think it makes sense to force users into writing additional tags since the whole purpose of our custom elements is to generate the html by ourselves when our users write simple things.
- `ons-repeat` - I think if we add a templating function maybe this won't be as necessary as the templates could take care of this, but I guess it's still an option. The fact that you're suggesting for it to have the same api as lazy-repeat would make the transition really easy which is good. IMO if looked at separately I think there could be a better api though. At least what I would envision would be something similar to:

  ```
<ons-repeat items="app.items" template="app.template">
	<ons-list-item>{{item.content}}</ons-list-item>
</ons-repeat>
```

  Where template would be needed only if it differs from the default one.
  
  But I think most of the times users would like to have a default template and just use their items array rather than generate fibonacci or something similar. So this is how I would prefer it, but I would understand if we decide to go with your original idea :)

- `ons.util.createElement` - actually I don't like this idea for multiple reasons, even though I seem to be the only one on the team who feels that way. 
  1. While the idea of exporting some utility functions sounds nice I think it would be better if we look into what they actually need, whereas here it sounds like we just want to export it since we're using it a lot internally. So maybe here @munsterlander can share what utility functions he would find useful.
  2. If we decide to export some functions I think they should be in `ons.util`, but they should be completely separate from our internal ones. We should be able to change the internal ones at any moment, while the public ones should not really change that much. By using the same ones we open ourselves to potential issues.
  3. I think this function in particular should not actually be used by the users. While in the core currently it may make sense, for the users I cannot think of many good reason to use it. 
  
    The only thing I can think of is making the elements in a loop and attaching event listeners to the children. However I think performance-wise that would not be very good. But other than that case when would actually creating a single element from html be beneficial and the right approach? If we want to use html we could just use innerHTML directly - we're not giving them better security than that because that's what we're doing under the hood. And if they're using it in a loop it's much better to concatenate the html for all elements and dump it at once rather than creating each element individually and appending them one by one (even if we're using fragments). 

  Basically I just think that if we give this feature to the users then the cases of improper usage would vastly outnumber the proper ones. 

Side note: @munsterlander - is there a problem with [`v-for`](https://vuejs.org/guide/list.html) ? Is it incompatible with onsen or is there a reason why it doesn't work for you currently?

**Comment by frankdiox** [comment]
@IliaSky Thanks for the feedback!
About `ons.util.createElement` part 3 `when would actually creating a single element from html be beneficial and the right approach?`. To-Do app, adding a new task could be like creating a single element that could be appended not only at the end of the list, but also somewhere at the middle (if it's for example an alphabetical ordered list). Maybe this is not the best example but I believe there could exist some usecase. Anyway, I don't really have a strong opinion on this since it's not that important.

**Comment by munsterlander** [comment]
So, to add some insight on how I would be using createElement, it comes down to variations - which to be honest, there are lots of ways to do it.  As @frankdiox mentioned, let's just use a to-do app.

You can use v-for in the view portion of your code, modify the dataset and refresh your view and that works great EXCEPT for when I want to do conditional templating.  Say my list is mixed.  Some images, maybe a form based on an image, or just text.  While you can do that with some creative view templating, it is much much easier to just do it in JS by dynamically creating your elements and then push them to the DOM.

Now, for ons-repeat, it is something that I would normally (and take normal with a slight grain of salt as I just started using vue.js ever since @argelius mentioned it) utilize v-for with; however, as my post history indicates, I am inherently against adding frameworks to frameworks due to  compatibility issues, etc.  So, if Onsen team decided to add ons-repeat, I would be happy to use it.  It is not a deal breaker for me, but it definitely has some significant advantages.

The same kind of goes for ons-include.  I am sure I could use it, but for the moment I am having difficulty coming up with how I would use to the fullest.  I would hope that I could define a template that has specific JS and that it would run via include.  If not, then I am not exactly certain how much of an impact it would make in my code (yes, I know I went all caps of excitement in my first post, I am running on caffeine fumes at the moment :o)

As far as templating goes, I do agree that any proper templating engine has some format to pass functions and  it should be added; having said that however, I have developed so many workarounds for it nowadays, I am not certain how desperate I am for it, but it is one that would make much cleaner code and provide a better MVC adherence.

Ok, I will cease my rambling.  Thanks again guys / gals for all the hard work that is being done on Onsen!
